adapters/tlsio_ssl_esp8266.c

```diff
@@ -0,0 +1,712 @@
+
+            result->x509certificate = NULL;
+            result->x509privatekey = NULL;
+        }
+    }
+
+    return (CONCRETE_IO_HANDLE)result;
+}
+
+/* Codes_SRS_TLSIO_SSL_ESP8266_99_010: [ The tlsio_openssl_destroy succeed ]*/
+void tlsio_openssl_destroy(CONCRETE_IO_HANDLE tls_io)
+{
+    /* Codes_SRS_TLSIO_SSL_ESP8266_99_009: [ The tlsio_openssl_destroy NULL parameter. make sure there is no crash ]*/
+    if (tls_io == NULL)
+    {
+        LogError("NULL tls_io.");
+    }
+    else
+    {
+        TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)tls_io;
+        if (tls_io_instance->certificate != NULL)
+        {
+            free(tls_io_instance->certificate);
+            tls_io_instance->certificate = NULL;
+        }
+        if (tls_io_instance->hostname != NULL)
+        {
+            free(tls_io_instance->hostname);
+        }
+        if (tls_io_instance->x509certificate != NULL)
+        {
+            free((void*)tls_io_instance->x509certificate);
+        }
+        if (tls_io_instance->x509privatekey != NULL)
+        {
+            free((void*)tls_io_instance->x509privatekey);
+        }
+        free(tls_io);
+    }
+}
+
+
+/* Codes_SRS_TLSIO_SSL_ESP8266_99_008: [ The tlsio_openssl_open succeed ]*/
+int tlsio_openssl_open(CONCRETE_IO_HANDLE tls_io, ON_IO_OPEN_COMPLETE on_io_open_complete, void* on_io_open_complete_context, ON_BYTES_RECEIVED on_bytes_received, void* on_bytes_received_context, ON_IO_ERROR on_io_error, void* on_io_error_context)
+{
+    int result;
+
+    if (tls_io == NULL)
+    {
+        /* Codes_SRS_TLSIO_SSL_ESP8266_99_006: [ The tlsio_openssl_open failed because tls_io is NULL. ]*/
+        result = __LINE__;
+        LogError("NULL tls_io.");
+    }
+    else
+    {
+
+        TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)tls_io;
+
+        /* Codes_SRS_TLSIO_SSL_ESP8266_99_007: [ The tlsio_openssl_open invalid state. ]*/
+        if (tls_io_instance->tlsio_state != TLSIO_STATE_NOT_OPEN)
+        {
+            result = __LINE__;
+            LogError("Invalid tlsio_state. Expected state is TLSIO_STATE_NOT_OPEN.");
+        }
+        else
+        {
+            tls_io_instance->on_io_open_complete = on_io_open_complete;
+            tls_io_instance->on_io_open_complete_context = on_io_open_complete_context;
+
+            tls_io_instance->on_bytes_received = on_bytes_received;
+            tls_io_instance->on_bytes_received_context = on_bytes_received_context;
+
+            tls_io_instance->on_io_error = on_io_error;
+            tls_io_instance->on_io_error_context = on_io_error_context;
+
+            tls_io_instance->tlsio_state = TLSIO_STATE_OPENING;
+
+            if (create_openssl_instance(tls_io_instance) != 0)
+            {
+
+                tls_io_instance->tlsio_state = TLSIO_STATE_ERROR;
+                result = __LINE__;
+            }
+            else
+            {
+                send_handshake_bytes(tls_io_instance);
```

Comment:
Please create the mechanism to propagate errors in the handshake to this point and check it to decide if this function shall return success of fail.